bpf: Add bind{4,6} programs to block NodePorts

gandro opened the pull request:
This attaches new programs to be called on bind{4,6} to block user
processes from binding ports that overlap with installed NodePort or
ExternalIP services. This ensures that BPF NodePort never accidentally
hijacks traffic to such processes. Sockets which were bound before the
BPF program is installed are not affected - this behavior is consistent
with kube-proxy.

**TODO**:

- [x] Missing CI tests
- [x] We probably want to use `post_bind{4,6}` instead, to make sure we check the actually assigned port (especially in case the user requested an ephemeral port). Requires iproute2 patch.

Fixes: #9206

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9880)
<!-- Reviewable:end -->

gandro committed:
bpf: Make sock_proto_enabled more generic

Changes the signature of sock_proto_enabled to directly take the
protocol as its argument. This allows it to be used with different
contexts such as `bpf_sock`.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -286,6 +286,49 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
+		 */
+		__u16 dport = bpf_ntohs(key.dport);
+		if (dport >= NODEPORT_PORT_MIN && dport <= NODEPORT_PORT_MAX &&
+			is_v4_loopback(key.address)) {
```
This case doesn't consider any other local address aside the two installed for nodeport. I would simply drop the `is_v4_loopback` to alway check if there is a 0.0.0.0/32 under nodeport port.

gandro replied:
Thanks for the review! I will adapt your suggestion, will also make the code simpler.

My original thinking was that some `bind()` calls on the nodeport port may not interfere with our NodePort forwarding. But blocking the nodeport port for all IPs will be consistent with kube-proxy, which also binds the port for `0.0.0.0`, so that should be fine.

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -286,6 +286,49 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
 }
 
+#ifdef ENABLE_NODEPORT
+__section("bind-sock4")
```
Needs the usual ifdef v4/v6.

borkmann replied:
Overall looks great to me!

gandro replied:
> Needs the usual ifdef v4/v6.

Not sure what you mean here - the `ifdef ENABLE_IPV4/6` are here (afaict):
https://github.com/cilium/cilium/blob/094c66962dd33c0a0c0b6a88ebe6765c45fcbb8d/bpf/bpf_sock.c#L114
https://github.com/cilium/cilium/blob/094c66962dd33c0a0c0b6a88ebe6765c45fcbb8d/bpf/bpf_sock.c#L421

borkmann replied:
Sigh, yes, missed that.

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -286,6 +286,49 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
+		return SYS_PROCEED;
+
+	svc = __lb4_lookup_service(&key);
+
```
(Tiny nit: no newline needed.)

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -286,6 +286,49 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
 	return SYS_PROCEED;
 }
 
+#ifdef ENABLE_NODEPORT
```
ENABLE_NODEPORT or ENABLE_EXTERNAL_IP

brb reviewed (commented):
> Sockets which were bound before the BPF program is installed are not affected - this behavior is consistent with kube-proxy.

What happens in this case? Do we allow provisioning of a svc which port clashes with an application? AFAIK, kube-proxy would log an error and would not provision the svc.

gandro committed:
bpf: Add post_bind{4,6} programs to block NodePorts

This attaches new programs to be called after bind{4,6} to block user
processes from binding ports that overlap with installed NodePort or
ExternalIP services. This ensures that BPF NodePort never accidentally
hijacks traffic to such processes. Sockets which were bound before the
BPF program is installed are not affected - this behavior is consistent
with kube-proxy.

Fixes: #9206

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
test: Add CI tests for blocked NodePorts

This tests that ports allocated to a NodePort service cannot not be
bound by local processes.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -577,6 +622,67 @@ static __always_inline int sock6_xlate_v4_in_v6(struct bpf_sock_addr *ctx)
+	}
+
+	if (svc && (lb6_svc_is_nodeport(svc) || lb6_svc_is_external_ip(svc))) {
+		return SYS_REJECT;
```
This is not fully correct; needs to be `return -EADDRINUSE` since otherwise `sock6_post_bind ()` will never trigger the `SYS_REJECT`. Lets also make the other return codes in here similar to the v4 variant of your code.

gandro replied:
:scream: Good catch!

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -595,12 +701,10 @@ static __always_inline int __sock6_xlate(struct bpf_sock_addr *ctx)
 	svc = __lb6_lookup_service(&key);
 	if (!svc) {
-		__builtin_memset(&key.address, 0, sizeof(key.address));
+		ctx_get_v6_address(ctx, &key.address);
```
Do we need to fetch this twice? The `key.address` should remain unchanged after the `__lb6_lookup_service()`; I think same should actually also be for the port.

gandro replied:
The port is actually reset, see: https://github.com/cilium/cilium/blob/b2c9b075ec92b0ed820c6f31cf358c268e95fff7/bpf/lib/lb.h#L682

But I'll remove the address.

borkmann replied:
Ah, yes, sounds good!

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -206,36 +206,37 @@ static __always_inline bool sock4_skip_xlate(struct lb4_service *svc,
-						   struct lb4_key *key)
+static __always_inline
+struct lb4_service *sock4_nodeport_wildcard_lookup(struct lb4_key *key,
+						   bool include_remote_hosts)
```
bool can be `const` so compiler can optimize when inlining.

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -319,6 +314,49 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
 	return SYS_PROCEED;
 }
 
+#if defined ENABLE_NODEPORT || defined ENABLE_EXTERNAL_IP
```
nit: `defined(ENABLE_NODEPORT) || defined(ENABLE_EXTERNAL_IP)`

gandro replied:
I will change it! Wasn't sure which style to use, since we seem to have two different styles in our code, I found a lot of code which doesn't have the parenthesis either (e.g. `bpf_lxc.c`).

borkmann replied:
Ah good point, might be good to clean this up at some point. If in doubt, always kernel style. :)

borkmann reviewed (approved):
LGTM, thanks!
